netperf.py

```diff
@@ -0,0 +1,125 @@
+from thread import ReturnValueThread
+from task import Task
+from common import Result
+from testSettings import TestSettings
+import json
+import time
+from syncManager import SyncManager
+import sys
+import perf
+
+NETPERF_SERVER_EXE = "netserver"
+NETPERF_CLIENT_EXE = "netperf"
+
+
+class NetPerfServer(perf.PerfServer):
+    def __init__(self, tc: TestConfig, ts: TestSettings):
+
+        self.exec_persistent = ts.server_is_persistent
+        if self.exec_persistent:
+            self.template_args["command"] = NETPERF_SERVER_EXE
+            self.template_args["args"] = f'["-p", "{self.port}", "-N"]'
+
+        perf.PerfServer.__init__(self, tc, ts)
+
+    def setup(self) -> None:
+        if self.connection_mode == ConnectionMode.EXTERNAL_IP:
+            cmd = f"podman run -it --rm -p {self.port} --entrypoint {NETPERF_SERVER_EXE} --name={self.pod_name} {common.TFT_TOOLS_IMG} -p {self.port} -N"
+            cleanup_cmd = f"podman rm --force {self.pod_name}"
+        else:
+            # Create the server pods
+            super().setup()
+            cmd = f"exec {self.pod_name} -- {NETPERF_SERVER_EXE} -p {self.port} -N"
+            cleanup_cmd = f"exec -t {self.pod_name} -- killall {NETPERF_SERVER_EXE}"
+
+        logger.info(f"Running {cmd}")
+
+        def server(self: NetPerfServer, cmd: str) -> Result:
+            if self.connection_mode == ConnectionMode.EXTERNAL_IP:
+                return self.lh.run(cmd)
+            elif self.exec_persistent:
+                return Result("Server is persistent.", "", 0)
+            return self.run_oc(cmd)
+
+        self.exec_thread = ReturnValueThread(
+            target=server,
+            args=(self, cmd),
+            cleanup_action=server,
+            cleanup_args=(self, cleanup_cmd),
+        )
+        self.exec_thread.start()
+        self.confirm_server_alive()
+
+
+class NetPerfClient(perf.PerfClient):
+    def __init__(self, tc: TestConfig, ts: TestSettings, server: NetPerfServer):
+        perf.PerfClient.__init__(self, tc, ts, server)
+
+    def run(self, duration: int) -> None:
+        def client(self: NetPerfClient, cmd: str) -> Result:
+            SyncManager.wait_on_barrier()
+            r = self.run_oc(cmd)
+            SyncManager.set_client_finished()
+            return r
+
+        server_ip = self.get_target_ip()
+        if self.test_type == TestType.NETPERF_TCP_STREAM:
+            self.cmd = f"exec {self.pod_name} -- {NETPERF_CLIENT_EXE} -H {server_ip} -p {self.port} -t TCP_STREAM -l {duration}"
+        else:
+            self.cmd = f"exec {self.pod_name} -- {NETPERF_CLIENT_EXE} -H {server_ip} -p {self.port} -t TCP_RR -l {duration}"
+
+        if self.reverse:
+            logger.info(f"Reverse is not supported by Netperf")
+
+        logger.info(f"Running {self.cmd}")
+
+        self.exec_thread = ReturnValueThread(target=client, args=(self, self.cmd))
+        self.exec_thread.start()
+
+    def generate_output(self, data: str) -> IperfOutput:
```

Comment:
IperfOutput -> PerfOutput

Reply:
We will need to refactor this a bit to handle different result outputs from iperf and netperf. We can do this later.